vault/resource_secrets_sync_association.go

```diff
@@ -89,48 +106,99 @@ func secretsSyncAssociationWrite(ctx context.Context, d *schema.ResourceData, me
 	}
 
 	log.Printf("[DEBUG] Writing association to %q", path)
-	resp, err := client.Logical().WriteWithContext(ctx, path, data)
+	_, err := client.Logical().WriteWithContext(ctx, path, data)
 	if err != nil {
 		return diag.Errorf("error setting secrets sync association %q: %s", path, err)
 	}
 	log.Printf("[DEBUG] Wrote association to %q", path)
 
-	// expect accessor to be provided from mount
-	accessor, err := getMountAccessor(ctx, d, meta)
-	if err != nil {
-		return diag.Errorf("could not obtain accessor from given mount; err=%s", err)
+	// ex: gh/dest/gh-dest-1/mount/kv/secret/token
+	// unique. each destination should only have one association for a particular accessor and secret
+	id := fmt.Sprintf("%s/dest/%s/mount/%s/secret/%s", destType, name, mount, secretName)
+	d.SetId(id)
+
+	return secretsSyncAssociationRead(ctx, d, meta)
+}
+
+func secretsSyncAssociationRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client, e := provider.GetClient(d, meta)
+	if e != nil {
+		return diag.FromErr(e)
 	}
-	vaultRespKey := fmt.Sprintf("%s/%s", accessor, secretName)
 
-	saModel, err := getSyncAssociationModelFromResponse(resp)
+	id := d.Id()
+	fields, err := syncAssociationFieldsFromID(id)
 	if err != nil {
 		return diag.FromErr(err)
 	}
 
-	syncData, ok := saModel.AssociatedSecrets[vaultRespKey]
-	if !ok {
-		return diag.Errorf("no associated secrets found for given mount accessor and secret name %s", vaultRespKey)
+	typ := fields[0]
+	destName := fields[1]
+	mount := fields[2]
+	secretName := fields[3]
+
+	if err := d.Set(fieldSecretName, secretName); err != nil {
+		return diag.FromErr(err)
 	}
 
-	// set data that is received from Vault upon writes to avoid extra sync association reads
-	if err := d.Set(fieldSecretName, syncData.SecretName); err != nil {
+	if err := d.Set(consts.FieldName, destName); err != nil {
 		return diag.FromErr(err)
 	}
 
-	if err := d.Set(fieldSyncStatus, syncData.SyncStatus); err != nil {
+	if err := d.Set(consts.FieldType, typ); err != nil {
 		return diag.FromErr(err)
 	}
 
-	if err := d.Set(fieldUpdatedAt, syncData.UpdatedAt); err != nil {
+	if err := d.Set(consts.FieldMount, mount); err != nil {
 		return diag.FromErr(err)
 	}
 
-	d.SetId(path)
+	accessor, err := getMountAccessor(ctx, d, meta, mount)
+	if err != nil {
+		return diag.Errorf("could not obtain accessor from given mount; err=%s", err)
+	}
+	// List all associations for secret destination
+	resp, err := client.Logical().Read(fmt.Sprintf("%s/%s", syncutil.SecretsSyncDestinationPath(destName, typ), "associations"))
+	if err != nil {
+		return diag.Errorf("error reading associations for destination %s of type %s", destName, typ)
+	}
 
-	return nil
-}
+	model, err := getSyncAssociationModelFromResponse(resp)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	status := map[string]string{}
```

Comment:
Since we're already storing the sync status for each subkey, would it make sense to also add the `updatedAt` field?

Reply:
@raymonstah that is doable, but do we see that as beneficial? My understanding was that the updated time should be the same for all subkeys, is that not the case? Do we expect to see a lot of cases where users have different sync update times for different subkeys?

Reply:
Today, they should share the same update times, because we sync the entire secret regardless of what changes were made. In future improvements, it's possible that individual subkey associations would have different update times. For example, if a new subkey were added to an existing secret, we may not need to sync the existing subkeys.

I'm mostly suggesting it to mimic Vault's structure of an association, and to avoid relying on the internal implementation details as that could change.

Reply:
Thanks for the explanation and insight into future directions, @raymonstah! That makes sense, I pushed the latest commit which also modifies the `updated_at` field to be a map, storing all update string information across all subkeys